gqcp/include/QCModel/HF/UHF.hpp

```diff
@@ -345,6 +347,656 @@ class UHF {
+     * 
+     *  @return The complete stability matrix B''.
+     */
+    GQCP::MatrixX<Scalar> calculateSpinUnconservedB(const RSQHamiltonian<Scalar>& rsq_hamiltonian) const {
+
+        // Calculate the two different B'' components.
+        const auto B_abba = this->calculateSpinUnconservedBComponent(rsq_hamiltonian, Spin::alpha, Spin::beta);  // Dimension = (n_occ_b * n_virt_a, n_occ_a * n_virt_b).
+        const auto B_baab = this->calculateSpinUnconservedBComponent(rsq_hamiltonian, Spin::beta, Spin::alpha);  // Dimension = (n_occ_a * n_virt_b, n_occ_b * n_virt_a).
+
+        // Determine the total matrix dimension and initialize the total matrix.
+        const auto n_occ_a = this->orbitalSpace(Spin::alpha).numberOfOrbitals(OccupationType::k_occupied);
+        const auto n_virt_a = this->orbitalSpace(Spin::alpha).numberOfOrbitals(OccupationType::k_virtual);
+        const auto n_occ_b = this->orbitalSpace(Spin::beta).numberOfOrbitals(OccupationType::k_occupied);
+        const auto n_virt_b = this->orbitalSpace(Spin::beta).numberOfOrbitals(OccupationType::k_virtual);
+
+        const auto dimension = (n_occ_b * n_virt_a) + (n_occ_a * n_virt_b);
+        GQCP::MatrixX<Scalar> total_B {dimension, dimension};
+
+        // Initialize the zero blocks of the total matrix.
+        const auto zero_1 = GQCP::MatrixX<Scalar>::Zero(n_virt_a * n_occ_b, n_occ_b * n_virt_a);
+        const auto zero_2 = GQCP::MatrixX<Scalar>::Zero(n_occ_a * n_virt_b, n_virt_b * n_occ_a);
+
+        // Place the components on the correct positions in the total matrix.
+        total_B.topLeftCorner(n_occ_b * n_virt_a, n_occ_b * n_virt_a) = zero_1;
+        total_B.topRightCorner(n_virt_a * n_occ_b, n_occ_a * n_virt_b) = B_abba;
+        total_B.bottomLeftCorner(n_occ_a * n_virt_b, n_occ_b * n_virt_a) = B_baab;
+        total_B.bottomRightCorner(n_occ_a * n_virt_b, n_occ_a * n_virt_b) = zero_2;
+
+        return total_B;
+    }
+
+
+    /**
+     *  Calculate the UHF stability matrices and return them.
+     *
+     *  @return The UHF stability matrices.
+     */
+    UHFStabilityMatrices<Scalar> calculateStabilityMatrices(const RSQHamiltonian<Scalar>& rsq_hamiltonian) const {
+        return UHFStabilityMatrices<Scalar> {this->calculateSpinConservedA(rsq_hamiltonian), this->calculateSpinConservedB(rsq_hamiltonian),
+                                             this->calculateSpinUnconservedA(rsq_hamiltonian), this->calculateSpinUnconservedB(rsq_hamiltonian)};
+    }
+
+
+    /**
+     *  @param sigma                The spin-sigma component. 
+     * 
+     *  @return A matrix containing all the possible excitation energies of the wavefunction model, belonging to a certain spin component. 
+     * 
+     *  @note       The rows are determined by the number of virtual sigma orbitals, the columns by the number of occupied sigma orbitals.
+     * 
+     */
+    GQCP::MatrixX<Scalar> excitationEnergies(const Spin sigma) const {
```

Comment:
For this method, `occupiedOrbitalEnergies` and `virtualOrbitalEnergies`,  could you maybe add a method that return a `SpinResolved<MatrixX<Scalar>>`, that holds the alpha and beta values together, analogously to what `orbitalSpace()` does?

You don't have to create a new class. We have a special helpful type `SpinResolved<T>` which encapsulates an alpha-`T` and a `beta`-T.